Add File Naming setting for more detailed naming

m1-ins opened the pull request:
# What's new

Solved Issue #2598 (Feature Request) by adding more detail to file names through the use of a non-invasive setting named 'File Naming'.

# Verification 

1. Go to: Settings > System > File Naming > Set to 'Detailed'
2. Create any file that is an internal app which supports the naming convention (Sub-Ghz, NFC, iButton, RFID)
3. Save the file

You can check the difference by setting the filename mode to 'Default'.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

m1-ins committed:
added filename mode setting

m1-ins committed:
added furi_flag checks for when filename_mode is set

m1-ins committed:
Merge branch 'flipperdevices:dev' into dev

m1-ins committed:
changed naming for ibutton, lfrfid and subghz

m1-ins committed:
Merge branch 'flipperdevices:dev' into dev

m1-ins committed:
Merge branch 'flipperdevices:dev' into dev

m1-ins committed:
Merge branch 'dev' into dev

skotopes commented on firmware/targets/furi_hal_include/furi_hal_rtc.h:
```diff
@@ -31,7 +31,8 @@ typedef enum {
     FuriHalRtcFlagHandOrient = (1 << 4),
     FuriHalRtcFlagLegacySleep = (1 << 5),
-    FuriHalRtcFlagStealthMode = (1 << 6),
+    FuriHalRtcFlagFilenameMode = (1 << 6),
```
Nope, you can not replace already existing flag

skotopes replied:
Also mode in FuriHalRtcFlagFilenameMode kinda expects that there will be some options. Not sure how it should be called.

m1-ins replied:
Leaving this open, as I am unsure if this change is visual / the right fix. I made sure to keep FuriHalRtcFlagStealthMode, but added FuriHalRtcFlagFilenameScheme in the most recent commit.

skotopes commented on lib/toolbox/set_name.c:
```diff
@@ -0,0 +1,19 @@
+#include <stdint.h>
+#include <furi_hal_rtc.h>
+
+void set_name(char* name, uint8_t max_name_size, char* app_name) {
```
I'd rather see one universal interface for random and detailed api with rtc flag check on the inside

skotopes commented on firmware/targets/f7/api_symbols.csv:
```diff
@@ -1,5 +1,5 @@
 entry,status,name,type,params
-Version,+,35.0,,
+Version,+,36.0,,
```
Major bump is not needed here

skotopes commented on applications/main/ibutton/scenes/ibutton_scene_save_name.c:
```diff
@@ -17,7 +19,11 @@ void ibutton_scene_save_name_on_enter(void* context) {
 
     if(is_new_file) {
-        set_random_name(ibutton->key_name, IBUTTON_KEY_NAME_SIZE);
+        if(furi_hal_rtc_is_flag_set(FuriHalRtcFlagFilenameMode)) {
```
It would be nice to see universal interface to generate names here

m1-ins committed:
requested changes from PR

m1-ins committed:
Merge branch 'dev' into dev

skotopes committed:
Merge branch 'dev' into dev

skotopes committed:
Lib: gather all naming bits and pieces under name generator module. Properly bump api version. FuriHal: fix RTC flag enum.

skotopes commented on lib/toolbox/name_generator.c:
```diff
@@ -0,0 +1,82 @@
+    };
+
+    struct NameConversion conversions[] = {
+        {".ibtn", "iBtn"}, {".nfc", "NFC"}, {".rfid", "RFID"}, {".sub", "RAW"}};
```
That doesn't looks extensible. Why not to pass it explicitly from application side?

skotopes commented on applications/main/subghz/scenes/subghz_scene_save_name.c:
```diff
@@ -40,7 +41,9 @@ void subghz_scene_save_name_on_enter(void* context) {
         char file_name_buf[SUBGHZ_MAX_LEN_NAME] = {0};
-        set_random_name(file_name_buf, SUBGHZ_MAX_LEN_NAME);
+
+        name_generator_set(file_name_buf, SUBGHZ_MAX_LEN_NAME, SUBGHZ_APP_EXTENSION);
```
That is not raw

skotopes commented:
Un-draft when ready

m1-ins committed:
Merge branch 'flipperdevices:dev' into dev

m1-ins committed:
PR requested changes

m1-ins committed:
bug fix for arg type

m1-ins committed:
added functionality for other application scenes

m1-ins committed:
Merge branch 'dev' into dev

m1-ins commented:
I believe I have addressed the issues raised in the last code review, but if there is logic/ the implementation that needs be changed specifically within the `convert_app_extension_to_name` function, let me know!

skotopes commented on lib/toolbox/name_generator.h:
```diff
@@ -0,0 +1,30 @@
+ * @param app_name name of the application
+ * @return converted name
+ */
+const char* convert_app_extension_to_name(char* app_name);
```
There is no need in this method. Application that uses name generator already knows its name, just provide there.

skotopes commented on lib/toolbox/name_generator.c:
```diff
@@ -0,0 +1,81 @@
+        {".ibtn", "iBtn"}, {".nfc", "NFC"}, {".rfid", "RFID"}, {".sub", "SubGhz"}};
+
+    const int numConversions =
+        sizeof(conversions) / sizeof(conversions[0]); // gets the number of potential conversions
```
COUNT_OF()

skotopes committed:
Merge branch 'dev' into dev

skotopes committed:
Lib: cleanup name generator API, simplify usage. Sync API symbols.

skotopes committed:
Lib: proper size type in name_generator. Cleanup.

skotopes committed:
FuriHal: cleanup rtc api usage across firmware
